### DISCUSSION TITLE
How to get number of constraints for a circuit?

### DISCUSSION THREAD

**Discussion by sshravan** [Consensys/gnark#115] [Q&A] [answered]
I am trying to find the total number of constraints for a circuit, say [this](https://github.com/ConsenSys/gnark/blob/master/examples/mimc/mimc.go).

Also, how to get stats after setup? Say like total number of variable, total of witness variables, number of G1 elements in PK, etc? 

Not sure if playing with the output of `frontend.Compile` helps.

**Comment by ThomasPiellard** [comment, marked as answer]
Hi,

you have access to the number of variables (public, secret, internal) and to the number of constraints through a compiled circuit like this:
```
type exp struct {
	X frontend.Variable
	Z frontend.Variable `gnark:",public"`
}

func (circuit *exp) Define(curveID ecc.ID, cs *frontend.ConstraintSystem) error {
	tmp := circuit.X
	for i := 0; i < 5; i++ {
		tmp = cs.Mul(tmp, circuit.X)
	}
	cs.AssertIsEqual(tmp, circuit.Z)
	return nil
}

func main() {

	var circuit exp
	ccircuit, err := frontend.Compile(ecc.BN254, backend.GROTH16, &circuit)
	if err != nil {
		fmt.Println(err)
		os.Exit(-1)
	}

	fmt.Printf("nb constraints: %d\n", ccircuit.GetNbConstraints())
	nbInternals, nbSecret, nbPublic := ccircuit.GetNbVariables()
	fmt.Printf("nb internal variables: %d\n", nbInternals)
	fmt.Printf("nb secret variables: %d\n", nbSecret)
	fmt.Printf("nb public variables: %d\n", nbPublic)

}
```

The number of elements in the keys are:

- verifying key: 3 + nb_public_wires G1 elmts, 5 G2 elements
- proving key:  3+ 3*(total_witness_wires) + nb_public_wires G1 elmts, 2 + total_witness_wires G2 elements

I can open an issue to add stats of this sort on the compiled circuit if you think this is helpful.

**Reply by sshravan** [reply]
Ah, thanks! 
Personally, I have found them useful! It would be great to have stats about PK and VK.

**Reply by spector-in-london** [reply]
That's a great idea.

**Reply by shreyas-londhe** [reply]
This was really helpful, thanks a lot
